crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,237 @@
+//! Different formatting functions.
+
+use std::time::Duration;
+
+/// Returns value with `n` digits after floating point where `n` is `precision`.
+/// Standard rounding rules apply (if `n+1`th digit >= 5, round up).
+///
+/// If rounding the `value` will have no effect (e.g., it's infinite or
+/// NaN), returns `value` unchanged.
+///
+/// # Examples
+///
+/// ```
+/// # use typst_utils::format::round_with_precision;
+/// let rounded = round_with_precision(-0.56553, 2);
+/// assert_eq!(-0.57, rounded);
+/// ```
+pub fn round_with_precision(value: f64, precision: u8) -> f64 {
+    // Don't attempt to round the float if that wouldn't have any effect. This
+    // includes infinite or NaN values, as well as integer values
+    // with a filled mantissa (which can't have a fractional part). Rounding
+    // with a precision larger than the amount of digits that can be
+    // effectively represented would also be a no-op.
+    if value.is_infinite()
+        || value.is_nan()
+        || value.abs() >= (1_i64 << f64::MANTISSA_DIGITS) as f64
+        || precision as u32 >= f64::DIGITS
+    {
+        return value;
+    }
+    let offset = 10_f64.powi(precision.into());
+    assert!((value * offset).is_finite(), "{value} * {offset} is not finite!");
```

Comment:
Since this is called from user code (`repr`), it'd be better to not have this panic in practice, being there just for debug purposes if we ever find this through testing:

```suggestion
    debug_assert!((value * offset).is_finite(), "{value} * {offset} is not finite!");
```

Reply:
But real testing only happens in production. How will we know if a user hits the assertions (that shouldn't be possible) if the assertion will be removed? Not everything must exist only during development or testing.

Fundamental assessments of the state of the program must always be true for it to work properly. If the assessment is false, then the state is corrupted and the logic is flawed.

Reply:
I understand, but displaying "infinity minutes" isn't fatal, just a bit ugly.

Reply:
Displaying — _maybe_, but this means that the function failed to do what it was called for. **That** is fatal.

<details><summary>When rounding function returns infinity instead of a rounded number:</summary>

![image](https://github.com/user-attachments/assets/4cbeff52-4d7c-4f4f-be11-ffdc08e1b3e8)

</details>

Reply:
I spoke with Laurenz about this, and we decided that the assertion should be removed entirely. Instead, we should ensure through the checks at the top that there won't be addition with overflow. (Some form of "formal verification", not literally but some manual validation of this, could be nice here.)

Reply:
We already [ensure it](https://github.com/typst/typst/blob/0bd3dd19108cae1dd481190325ea3d1b4f8919d5/crates/typst-utils/src/format.rs#L24-L27). The assertion is just a sanity check.

I don't understand the verification/validation part.

Reply:
I meant ensure that it is logically impossible for there to be overflow. Then we should remove the assertion. Laurenz said that any possibility of user code panicking is a bug.

Reply:
> Laurenz said that any possibility of user code panicking is a bug.

Well, duh, assertion provides the ability to know what the bug specifically is. It's almost the exact same thing as "this is a bug" error that was present not so long ago that I hit.

> it is logically impossible for there to be overflow.

It _is_ logically impossible, the assertion is _just_ a _sanity check_. I can remove it, if it's unacceptable. I'd rather have it than not. OK, I think we can agree on debug-only assertion then.

Reply:
If it is logically impossible it's fine.

Reply:
You mean the `assert!()`?

Reply:
Yes